File: src/snake.erl
-module(snake).
-behavior(gen_server).

%% TODO check for errors (mostly move/collision part)!

%% API for snake module
-export([start_link/0, start_link/1, start_link/3]).

%% gen server callbacks

-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
        terminate/2, code_change/3]).

-define(SERVER, ?MODULE).
-define(SUPERVISOR, 'Elixir.Snake.SnakeSupervisor').
-define(GUI, 'Elixir.Phoenix.Channel').

-define(WIDTH, 10). % TODO refactor this later to board module.
-define(HEIGHT, 10). % TODO refactor this later to board module.
-define(SCORE, 10). % TODO refactor this also (insect module?).

-record(state, {segment = head, 
                position = {0, 0}, 
                next_segment = no_pid,
                node = node()}).

%% API

%% Starts a snake process (head) at location {0, 0}.
start_link() -> start_link(head, 0, 0).

%% Starts a snake process (head or tail) at location {X, Y}.
start_link(head, X, Y) -> 
    Args = [head, X, Y],
    Options = [],
    gen_server:start_link(?MODULE, Args, Options);
start_link(tail, X, Y) ->
    Args = [tail, X, Y],
    Options = [],
    gen_server:start_link(?MODULE, Args, Options).

%% Starts a snake process with a certain state 
%% (only used by supervisor to move snake from node to node).
start_link([State = #state{}]) ->
    Args = [State#state{node = node()}], % Old state still contained old node?
    Options = [],
    gen_server:start_link(?MODULE, Args, Options).

%% gen server callbacks

init([head, X, Y]) ->
    Snake = self(),
    ticker:subscribe(fun(tick) -> gen_server:cast(Snake, update_state) end),
    State = #state{segment = head, position = {X, Y}},
    {ok, State};
init([tail, X, Y]) ->
    State = #state{segment = tail, position = {X, Y}},
    {ok, State};
init([State = #state{segment = head}]) ->
    Snake = self(),
    ticker:subscribe(fun(tick) -> gen_server:cast(Snake, update_state) end),   
    {ok, State};
init([State = #state{segment = tail}]) ->
    {ok, State}.

handle_call({is_collision, {X, Y, Node}}, _From, 
            State = #state{position = {X, Y}, node = Node}) ->
    % Same coordinates and node => collision!
    Reply = {is_collision, true},
    {reply, Reply, State};
handle_call({is_collision, {X, Y, Node}}, _From,
            State = #state{position = {X, Y},
                           next_segment = Next,
                           node = _OtherNode}) ->
    % Same coordinates, different nodes => no collision (here), check rest!
    Reply = is_collision(Next, X, Y, Node),
    {reply, Reply, State};
handle_call({is_collision, {X, Y, Node}}, _From, 
            State = #state{position = {_OtherX, _OtherY}, 
                           next_segment = Next}) ->
    % Different coordinates, different nodes => no collision (here), check rest!
    Reply = is_collision(Next, X, Y, Node),
    {reply, Reply, State};
handle_call({move, NewPosition, Node}, _From, 
            State = #state{next_segment = no_pid, 
                           node = Node}) ->
    % Only called in tail segments.
    % Previous part is on same node as this one, no special action needed.
    NewState = State#state{position = NewPosition},
    Reply = {tail, self()}, % pid stays the same => return self()!
    {reply, Reply, NewState};
handle_call({move, NewPosition, NewNode}, _From,
            State = #state{next_segment = no_pid}) ->
    % Only called in tail segments
    % Previous part is on another node => special action needed.
    % 1) Update state.
    NewState = State#state{position = NewPosition},
    % 2) Start the process on the new node with this new state.
    {ok, NewSegment} = ?SUPERVISOR:start_child(NewNode, NewState),
    % 3) Notify previous segment that this process is now on another node.
    % And finally stop this process.
    Reason = moving_to_other_node, 
    Reply = {tail, NewSegment},
    {stop, Reason, Reply, State};
handle_call({move, NewPosition, Node}, _From, 
            State = #state{position = OldPosition, 
                           next_segment = NextPid,
                           node = Node}) ->
    % Only called in tail segments.
    % Previous part is on same node as this one, no special action needed
    {tail, Segment} = gen_server:call(NextPid, {move, OldPosition, Node}),
    NewState = State#state{position = NewPosition, next_segment = Segment},
    Reply = {tail, self()},
    {reply, Reply, NewState};
handle_call({move, NewPosition, NewNode}, _From, 
            State = #state{position = OldPosition, 
                           next_segment = Next}) ->
    % Only called in tail segments
    % Previous part is on another node => special action needed.
    % 1) Notify next segment of movement
    {tail, Segment} = gen_server:call(Next, {move, OldPosition, node()}),
    % 2) Update state
    NewState = State#state{position = NewPosition, next_segment = Segment},
    % 3) Start the process on the new node with this new state.
    {ok, NewSegment} = ?SUPERVISOR:start_child(NewNode, NewState),
    % 4) Update previous segment that this segment is now on another node.
    % And finally stop this process.
    Reason = moving_to_other_node,
    Reply = {tail, NewSegment},
    {stop, Reason, Reply, NewState};
handle_call(_Request, _From, State) ->
    Reply = {error, not_supported},
    {reply, Reply, State}.

handle_cast(update_state, State = #state{}) ->
    % Only triggered in the head! => updates the entire snake recursively!
    % Ask for new input.
    Direction = input:get_direction(),
    
    % Move the snake.
    Result = move(Direction, State),
    handle_move(Result),
    Result; % TODO try merging these 2 handle_casts back together!
handle_cast(update_state2, State = #state{position = {NewX, NewY}, 
                                          next_segment = Next,
                                          node = Node}) ->
    % Check for collisions with itself
    {is_collision, SnakeCollision} = is_collision(Next, NewX, NewY, Node),
    handle_collision(SnakeCollision),

    % Check for collisions with insect => if yes: update score
    InsectLocation = insect:get_location(),
    {is_collision, InsectCollision} = is_insect_collision(InsectLocation, 
                                                          State),
    update_score(?SCORE, InsectCollision),

    % Draw the snake.
    draw(State),
    {noreply, State};
handle_cast(draw, State = #state{}) ->
    draw(State),
    {noreply, State};
handle_cast(gameover, State = #state{next_segment = no_pid}) ->
    {stop, gameover, State};
handle_cast(gameover, State = #state{next_segment = Next}) ->
    gen_server:cast(Next, gameover),
    {stop, gameover, State};
handle_cast(_Request, State) ->
    {noreply, State}.

handle_info(_Request, State) ->
    {noreply, State}.

terminate(_Reason, #state{segment = head}) -> 
    ticker:unsubscribe(),
    ok;
terminate(_Reason, _State = #state{}) ->
    ok.

code_change(_OldVsn, State, _Extra) -> {ok, State}.

%% Helper functions

%% GUI related
draw(#state{position = {X, Y}, next_segment = no_pid}) -> 
    draw_segment(X, Y);
draw(#state{position = {X, Y}, next_segment = NextPid}) ->
    draw_segment(X, Y),
    gen_server:cast(NextPid, draw). % TODO make this a call?

draw_segment(X, Y) -> ?GUI:broadcast("draw_snake", {X, Y}).

update_score(_Score, no_collision) -> ok;
update_score(Score, collision) -> do_update_score(Score).

do_update_score(Score) ->  ?GUI:broadcast("score", Score).

%% Other helpers

% Compares insect location with head of snake location:
is_insect_collision({X, Y, Node}, #state{position = {X, Y}, node = Node}) ->
    collision;
is_insect_collision(_InsectLocation, _SnakeState) ->
    no_collision.

move(Direction, State = #state{next_segment = no_pid}) -> 
    do_move(Direction, State);
move(Direction, State = #state{position = {OldX, OldY}, 
                               next_segment = Next,
                               node = Node}) ->
    % next function has to be synchronous, 
    % otherwise collision logic might be buggy.
    {tail, Segment} = gen_server:call(Next, {move, {OldX, OldY}, Node}),    
    NewState = State#state{next_segment = Segment},
    do_move(Direction, NewState).

do_move(left, State = #state{position = {X, Y}}) when X > 0 ->
    NewState = State#state{position = {X - 1, Y}},
    {noreply, NewState};
do_move(left, State = #state{position = {_X, Y}}) ->
    {board, NewNode} = board:get(left),
    NewState = State#state{position = {?WIDTH, Y}},
    NewHead = spawn_head(NewNode, NewState),
    Reason = {moving_to_other_node, NewHead},
    {stop, Reason, State};
do_move(right, State = #state{position = {X, Y}}) when X < ?WIDTH ->
    NewState = State#state{position = {X + 1, Y}},
    {noreply, NewState};
do_move(right, State = #state{position = {_X, Y}}) ->
    {board, NewNode} = board:get(right),
    NewState = State#state{position = {0, Y}},
    NewHead = spawn_head(NewNode, NewState),
    Reason = {moving_to_other_node, NewHead},
    {stop, Reason, State};
do_move(up, State = #state{position = {X, Y}}) when Y < ?HEIGHT ->
    NewState = State#state{position = {X, Y + 1}},
    {noreply, NewState};
do_move(up, State = #state{position = {X, _Y}}) ->
    {board, NewNode} = board:get(up),
    NewState = State#state{position = {X, 0}},
    NewHead = spawn_head(NewNode, NewState),
    Reason = {moving_to_other_node, NewHead},
    {stop, Reason, State};
do_move(down, State = #state{position = {X, Y}}) when Y > 0 ->
    NewState = State#state{position = {X, Y - 1}},
    {noreply, NewState};
do_move(down, State = #state{position = {X, Y}}) ->
    {board, NewNode} = board:get(down),
    NewState = State#state{position = {X, Y - 1}},
    NewHead = spawn_head(NewNode, NewState),
    Reason = {moving_to_other_node, NewHead},
    {stop, Reason, State}.

spawn_head(no_board, _State) -> 
    gen_server:cast(self(), gameover); % TODO make this a call, otherwise the {stop, ...} prevents game from cleaning up?
spawn_head(Node, State) -> 
    {ok, Head} = ?SUPERVISOR:start_child(Node, State),
    Head.

handle_move({noreply, _State}) ->
    gen_server:cast(self, update_state2);
handle_move({stop, {_Reason, NewHead}, _State}) ->
    gen_server:cast(NewHead, update_state2).

is_collision(no_pid, _X, _Y, _Node) -> {is_collision, false};
is_collision(NextPid, X, Y, Node) -> 
    % TODO maybe use cast here later? 
    % splits main game logic up in 2 parts though..
    gen_server:call(NextPid, {is_collision, {X, Y, Node}}).

handle_collision(false) -> ok;
handle_collision(true) -> gen_server:cast(self(), gameover).

File: lib/snake.ex
defmodule Snake.Snake do
  use GenServer
  use Snake.Game
  alias Snake.Ticker
  alias Snake.Insect
  alias Snake.Board
  alias Snake.Tile
  alias Snake.TileDB
  alias Snake.Random
  alias Snake.GUI

  @moduledoc """
  Module containing logic for the snake process.

  TODO:
  =====
  - gameover => simply crash
  - (re)start game + reacting to player input
  """

  @server {:global, __MODULE__} # TODO or local registered process?

  defmodule State do
    use Snake.Game

    @moduledoc """
    Module for the state of the snake. Also contains a few helper functions.
    """

    @default_value quote do: throw "Unassigned field in struct!"
    defstruct size: @default_value,
              positions: @default_value,
              previous_positions: @default_value,
              direction: @default_value,
              current_direction: @default_value,
              color: @default_value
    
    @doc """
    Creates a snake of size 'size' at a random {x, y} on a random node.
    """
    def new(size) do
      Random.generate_seed

      # First x and y somewhere in the middle of the board.
      first_x = round((Random.number(4) - 2) + 0.5 * width)
      first_y = round((Random.number(4) - 2) + 0.5 * height)
      first_node = Random.element [Node.self | Node.list]
      direction = Random.element [:up, :down, :left, :right]
      color = Random.element [:orange, :green, :blue, :cyan, 
                                :yellow, :purple, :white]

      tail = calc_tail_positions(first_x, first_y, first_node, 
                                  direction, size - 1, [])
      positions = [{first_x, first_y, first_node} | tail]

     state = %State{size: size, 
                    positions: positions, previous_positions: positions,
                    direction: direction, current_direction: direction,
                    color: color}
      
                  # Check if insect exists, prevent collisions
      Insect 
      |> :global.whereis_name 
      |> check_for_insect_collisions(state)
    end

    defp check_for_insect_collisions(:undefined, state), do: state
    defp check_for_insect_collisions(_insect, state) do
      {insect_x, insect_y, insect_node} = Insect.get_position
      state.positions |> Enum.filter(fn({x, y, n}) ->
        x == insect_x && y == insect_y && n == insect_node
      end)
      |> handle_collision_result(state)
    end

    defp handle_collision_result([], state), do: state
    defp handle_collision_result(_, state),  do: new(state.size)

    @doc """
    Moves the snake.
    """
    def move(state = %State{positions: [{x, y, a_node} | _tail], 
                            direction: :left}) when x > 0 do
      state 
      |> update_current_direction
      |> do_move(x - 1, y, {Board, a_node})
    end
    def move(state = %State{positions: [{_x, y, _node} | _tail], 
                            direction: :left}) do
      board = Board.get(:left)
      state
      |> update_current_direction
      |> do_move(width, y, board)
    end
    def move(state = %State{positions: [{x, y, a_node} | _tail], 
                            direction: :right}) when x < width - 1 do
      state
      |> update_current_direction
      |> do_move(x + 1, y, {Board, a_node})
    end
    def move(state = %State{positions: [{_x, y, _node} | _tail], 
                            direction: :right}) do
      board = Board.get(:right)
      state
      |> update_current_direction
      |> do_move(0, y, board)
    end
    def move(state = %State{positions: [{x, y, a_node} | _tail], 
                            direction: :up}) when y < height - 1 do
      state
      |> update_current_direction
      |> do_move(x, y + 1, {Board, a_node})
    end
    def move(state = %State{positions: [{x, _y, _node} | _tail], 
                            direction: :up}) do
      board = Board.get(:up)
      state 
      |> update_current_direction
      |> do_move(x, 0, board)
    end
    def move(state = %State{positions: [{x, y, a_node} | _tail], 
                            direction: :down}) when y > 0 do
      state 
      |> update_current_direction
      |> do_move(x, y - 1, {Board,a_node})
    end
    def move(state = %State{positions: [{x, _y, _node} | _tail], 
                            direction: :down}) do
      board = Board.get(:down)
      state 
      |> update_current_direction
      |> do_move(x, height, board)
    end

    @doc """
    Appends a segment to the end of the snake (same location as previous 
    last segment).
    """
    def add_segment(state = %State{positions: positions}) do
      reverse_pos = Enum.reverse positions
      new_positions = Enum.reverse [hd(reverse_pos) | reverse_pos]
      %State{state | size: state.size + 1, positions: new_positions}
    end

    # Helper functions

    defp update_current_direction(state = %State{direction: dir}) do
      %State{state | current_direction: dir}
    end

    defp do_move(state = %State{}, _x, _y, :no_board), do: {:gameover, state}
    defp do_move(state = %State{positions: pos}, new_x, new_y, 
                  {Board, new_node}) do
      # Drops last segment, 
      # adds new segment to beginning (new location for head)
      # and updates the state.
      {:ok, %State{state 
             | positions: [{new_x, new_y, new_node} | Enum.drop(pos, -1)],
              previous_positions: pos}}
    end

    defp calc_tail_positions(_x, _y, _node, _dir, 0, acc) do
      Enum.reverse acc
    end
    defp calc_tail_positions(x, y, a_node, :left, amount, acc) do
      calc_tail_positions(x + 1, y, a_node, :left, amount - 1, 
                          [{x + 1, y, a_node} | acc])  
    end
    defp calc_tail_positions(x, y, a_node, :right, amount, acc) do
      calc_tail_positions(x - 1, y, a_node, :right, amount - 1,
                          [{x - 1, y, a_node} | acc])
    end
    defp calc_tail_positions(x, y, a_node, :up, amount, acc) do
      calc_tail_positions(x, y - 1, a_node, :up, amount - 1,
                          [{x, y - 1, a_node}| acc])
    end
    defp calc_tail_positions(x, y, a_node, :down, amount, acc) do
      calc_tail_positions(x, y + 1, a_node, :down, amount - 1,
                          [{x, y + 1, a_node} | acc])
    end
  end


  # Snake API

  @doc """
  Starts the snake process (size = 4 blocks) at a random location {x,y}.
  """
  def start_link, do: start_link(3)

  @doc """
  Starts a snake of a certain size at a random location {x, y}.
  """
  def start_link(size) when is_integer(size) and size > 0 do
    GenServer.start_link(__MODULE__, State.new(size), [name: @server])
  end

  @doc """
  Returns a list of positions for each segment of the snake.
  """
  def get_positions do
    @server |> GenServer.call :get_positions
  end

  @doc """
  Updates the direction the snake will move to next.
  """
  def set_direction("up"),    do: update_direction(:up)
  def set_direction("down"),  do: update_direction(:down)
  def set_direction("left"),  do: update_direction(:left)
  def set_direction("right"), do: update_direction(:right)
 
  @doc """
  Subscribes the snake to the ticker process.
  """
  def subscribe_to_ticker do
    snake = :global.whereis_name __MODULE__
    Ticker.subscribe snake, fn(:tick) ->
      snake |> update_state
    end
  end

  @doc """
  Force the snake to redraw its current state
  """
  def draw do
    @server |> GenServer.call :draw
  end

  # GenServer callbacks

  @doc false
  def init(state = %State{positions: positions, color: color}) do
    positions |> Enum.map fn({x, y, a_node}) ->
      tile_notify_arrived(x, y, a_node, color)
    end

    # TODO also subscribe to insect for events!
    {:ok, state}
  end

  @doc false
  def handle_call(:get_positions, _from, state = %State{positions: pos}) do
    {:reply, pos, state}
  end
  def handle_call(:update_state, _from, state = %State{}) do
    {status, new_state} = 
      state
      |> State.move                   # 1) Move the snake
      |> check_for_snake_collisions   # 2) Check for collisions with itself
      |> check_for_insect_collisions  # 3) Check for collisions with insect
      |> draw_snake(state)            # 4) Update GUI

    {status, new_state} |> handle_status
    # TODO handle gameover better? TODO make supervisors permanent for board, game and tile!
  end
  def handle_call({:update_direction, :left}, _from, 
                  state = %State{current_direction: :right}) do
                    # Do not allow this (would instantly cause gameover),
                    # same for other clauses..
    {:reply, :ok, state}
  end
  def handle_call({:update_direction, :right}, _from, 
                  state = %State{current_direction: :left}) do
    {:reply, :ok, state}
  end
  def handle_call({:update_direction, :up}, _from, 
                    state = %State{current_direction: :down}) do
    {:reply, :ok, state}
  end
  def handle_call({:update_direction, :down}, _from, 
                  state = %State{current_direction: :up}) do
    {:reply, :ok, state}
  end
  def handle_call({:update_direction, new_dir}, _from, state = %State{}) do
    {:reply, :ok, %State{state | direction: new_dir}}
  end
  def handle_call(:draw, _from, state = %State{positions: pos, color: c}) do
    pos |> Enum.map fn({x, y, n}) ->
      GUI.draw_snake %{x: x, y: y, node: n, color: c}
    end
    {:reply, :ok, state}
  end
  def handle_call(_request, _from, state = %State{}) do
    {:reply, {:error, :not_supported}, state}
  end

  @doc false
  def terminate(_reason, %State{positions: positions, 
                                previous_positions: old_positions}) do
    Ticker.unsubscribe self
    insect = :global.whereis_name Insect
    Ticker.unsubscribe insect
    
    positions |> Enum.map fn({x, y, a_node}) ->
      tile_notify_gone(x, y, a_node)
    end

    old_positions |> Enum.map fn({x, y, a_node}) ->
      tile_notify_gone(x, y, a_node)
    end

    GUI.reset_score
    :ok
  end
  def terminate(_reason, state) do
    IO.inspect state
    :ok
  end


  # Helper functions

  defp check_for_snake_collisions(arg = {:gameover, _}), do: arg
  defp check_for_snake_collisions({:ok, 
                                  state = %State{positions: [head | tail]}}) do
    # Checks for collisions with itself
    handle_collision_result(head in tail, state)
  end

  defp handle_collision_result(false, state), do: {:ok, state}
  defp handle_collision_result(true, state), do:  {:gameover, state}

  defp check_for_insect_collisions(arg = {:gameover, _}), do: arg
  defp check_for_insect_collisions({:ok, 
                                    state = %State{positions: [head | _]}}) do
    handle_insect_collision_result(head, Insect.get_position, state)
  end
  
  defp handle_insect_collision_result(position, position, 
                                      state = %State{positions: positions}) do
    Insect.kill(positions) #  TODO refactor later (use agent!)
    GUI.update_score %{score: score}
    {:ok, state |> State.add_segment}
  end
  defp handle_insect_collision_result(_position1, _position2, state) do
    {:ok, state}
  end

  defp draw_snake(arg = {:gameover, _}, _state), do: arg
  defp draw_snake({:ok, new_state = %State{positions: new_positions, 
                                            color: color}}, 
                 _old_state = %State{positions: old_positions}) do
    # Only update first and last part of snake
    [{newest_x, newest_y, newest_node} | _] = new_positions
    [{last_x, last_y, last_node} |_] = Enum.reverse(old_positions)

    tile_notify_gone(last_x, last_y, last_node)
    tile_notify_arrived(newest_x, newest_y, newest_node, color)

    {:ok, new_state}
  end

  defp handle_status({:gameover, new_state}) do
    {:stop, :gameover, :ok, new_state}
  end
  defp handle_status({:ok, new_state}) do
    {:reply, :ok, new_state}
  end

  defp tile_notify_arrived(x, y, a_node, color) do
    snake = self
    Node.spawn a_node, fn ->
      {x, y}
      |> TileDB.get
      |> Tile.notify_arrival(:snake, snake, color)
    end
  end

  defp tile_notify_gone(x, y, a_node) do
    snake = self
    Node.spawn a_node, fn ->
      {x, y}
      |> TileDB.get
      |> Tile.notify_gone(snake)
    end
  end

  defp update_state(snake), do: :ok = snake |> GenServer.call :update_state
  defp update_direction(direction) do
    @server |> GenServer.call {:update_direction, direction}
  end
end
